pgmpy/factors/Factor.py

```diff
@@ -365,6 +379,14 @@ def reduce(self, values, inplace=True):
         >>> phi.values
         array([0., 1.])
         """
+
+        if isinstance(values, str):
+            raise TypeError("values: Expected type list or array-like, got type str")
+
+        if any(isinstance(value, str) for value in values) or \
+               not all(isinstance(var, str) and isinstance(state, int) for var,state in values):
+            raise TypeError("values: must contain tuples or array-like elements of the form (type str, type int)")
```

Comment:
@yashu-seth This is wrong. Variable names can be any hashable python object not just str. Please fix this.